File: packages/php/src/array_first.php
<?php

/**
 * Get the first element or first n elements of an array.
 * PHP port of https://github.com/mattstrick/array-first-javascript
 *
 * @param array $arr
 * @param int|null $n
 * @return mixed First element, first n elements as array, or null if empty
 */
function array_first(array $arr, ?int $n = null)
{
    if (count($arr) === 0) {
        return null;
    }
    if ($n === null) {
        return $arr[array_key_first($arr)];
    }
    $n = $n < 1 ? 1 : $n;
    $slice = array_slice($arr, 0, $n, true);
    return $n === 1 ? $slice[array_key_first($slice)] : array_values($slice);
}

File: packages/ruby/lib/array_first.rb
# frozen_string_literal: true

# Get the first element or first n elements of an array.
# Ruby port of https://github.com/mattstrick/array-first-javascript
module ArrayFirst
  module_function

  # Returns the first element or first n elements.
  # first(arr) => first element or nil
  # first(arr, n) => first n elements (single element if n is 1)
  def first(arr, n = nil)
    raise TypeError, "array-first expects an array as the first argument." unless arr.is_a?(Array)
    return nil if arr.empty?

    if n.nil?
      arr[0]
    else
      take = (n.is_a?(Integer) && n >= 1) ? n : 1
      slice = arr.first(take)
      take == 1 ? slice[0] : slice
    end
  end
end
